File: app/src/main/java/com/innoveworkshop/gametest/engine/GameObject.java
package com.innoveworkshop.gametest.engine;

import android.graphics.Canvas;
import android.view.SurfaceView;

public abstract class GameObject {
    public Vector position;

    protected GameSurface gameSurface = null;
    protected boolean destroyed = false;

    public GameObject() {
        this(0, 0);
    }

    public GameObject(Vector position) {
        this.position = position;
    }

    public GameObject(float x, float y) {
        position = new Vector(x, y);
    }

    public void setPosition(float x, float y) {
        position.x = x;
        position.y = y;
    }

    public void destroy() {
        gameSurface.removeGameObject(this);
        destroyed = true;
    }

    public boolean isDestroyed() {
        return destroyed;
    }

    public void onStart(GameSurface surface) {
        this.gameSurface = surface;
    }

    public void onFixedUpdate() {
        if (destroyed) {
            setPosition(-100, -100);
        }
    }

    public void onDraw(Canvas canvas) {}
}

File: app/src/main/java/com/innoveworkshop/gametest/engine/GameObject.kt
package com.innoveworkshop.gametest.engine

import android.graphics.Canvas

abstract class GameObject {
    var position: Vector

    protected var gameSurface: GameSurface? = null
    var bounce: Boolean = false
        protected set

    constructor(position: Vector) {
        this.position = position
    }

    @JvmOverloads
    constructor(x: Float = 0f, y: Float = 0f) {
        position = Vector(x, y)
    }

    fun setPosition(x: Float, y: Float) {
        position.x = x
        position.y = y
    }

    fun bounce() {
        //gameSurface!!.removeGameObject(this)
        bounce = true
    }

    open fun onStart(surface: GameSurface?) {
        this.gameSurface = surface
    }

    open fun onFixedUpdate() {
        if (bounce) {
            setPosition(100f, 100f)
        }
    }

    open fun onDraw(canvas: Canvas?) {}
}
